server/src/com/thoughtworks/go/server/websocket/ConsoleLogEndpoint.java

```diff
@@ -0,0 +1,87 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.thoughtworks.go.server.websocket;
+
+import com.thoughtworks.go.domain.JobIdentifier;
+import com.thoughtworks.go.server.service.RestfulService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.websocket.EndpointConfig;
+import javax.websocket.OnError;
+import javax.websocket.OnOpen;
+import javax.websocket.Session;
+import javax.websocket.server.PathParam;
+import javax.websocket.server.ServerEndpoint;
+import java.io.IOException;
+
+@ServerEndpoint(
+        value = "/client-websocket/{pipelineName}/{pipelineLabel}/{stageName}/{stageCounter}/{jobName}",
+        configurator = ConsoleLogEndpointConfigurator.class
+)
+public class ConsoleLogEndpoint {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConsoleLogEndpoint.class);
+    private Session session;
+
+    @OnOpen
+    public void onOpen(Session session, EndpointConfig config,
+                       @PathParam(value = "pipelineName") String pipelineName,
+                       @PathParam(value = "pipelineLabel") String pipelineLabel,
+                       @PathParam(value = "stageName") String stageName,
+                       @PathParam(value = "stageCounter") String stageCounter,
+                       @PathParam(value = "jobName") String jobName) throws Exception {
+        this.session = session;
+        LOGGER.debug("{} connected.", this);
+
+        Long start = (Long) config.getUserProperties().get("startLine");
+        ConsoleLogSender sender = (ConsoleLogSender) config.getUserProperties().get("sender");
+        JobIdentifier jobIdentifier = getJobIdentifier(config, pipelineName, pipelineLabel, stageName, stageCounter, jobName);
+
+        LOGGER.info("Sending logs for {} starting at line {}", jobIdentifier, start);
+        sender.process(this, jobIdentifier, start);
+    }
+
+    public void send(String msg) throws IOException {
+        LOGGER.debug("{} send message: {}", this, msg);
+        session.getBasicRemote().sendText(msg);
+    }
+
+    public void close() {
+        LOGGER.debug("{} closing session.", this);
+        try {
+            session.close();
+        } catch (IOException e) {
+            LOGGER.warn("{} failed to close session.", this);
+        }
+    }
+
+    @OnError
+    public void onError(Session session, Throwable error) {
+        LOGGER.error(toString() + " closing session because an error was thrown", error);
```

Comment:
Sure, I think I wasn't sure whether or not it would treat the args like it does for debug and info; I didn't want the the throwable to be just thr.toString(), and I saw that there's a signature for `Logger.error(String message, Throwable throwable)` which I know does the right thing. If Logger.error(String message, Object... args) knows how to chop that last arg as a throwable and append the stacktrace instead, I'll change it.